File: sdk/java/src/main/java/com/alp/sdk/AlpGraph.java
package com.alp.sdk;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

public class AlpGraph {
    private final Map<String, GraphNode> nodes = new HashMap<>();
    private final Map<String, List<String>> adjacency = new HashMap<>();

    public void buildGraph(List<AlpObject> objects) {
        nodes.clear();
        adjacency.clear();
        for (AlpObject obj : objects) {
            nodes.put(obj.getId(), new GraphNode(obj.getId(), obj.getType()));
        }
        for (AlpObject obj : objects) {
            Object dependsOn = obj.getProperties().get("depends_on");
            if (dependsOn instanceof String depId) {
                adjacency.computeIfAbsent(depId, k -> new ArrayList<>()).add(obj.getId());
            }
        }
    }

    public List<GraphNode> topologicalSort() {
        Map<String, Integer> inDegree = new HashMap<>();
        for (String nodeId : nodes.keySet()) {
            inDegree.putIfAbsent(nodeId, 0);
        }
        for (List<String> deps : adjacency.values()) {
            for (String dep : deps) {
                inDegree.merge(dep, 1, Integer::sum);
            }
        }
        LinkedList<String> queue = new LinkedList<>();
        for (Map.Entry<String, Integer> entry : inDegree.entrySet()) {
            if (entry.getValue() == 0) {
                queue.add(entry.getKey());
            }
        }
        List<GraphNode> result = new ArrayList<>();
        while (!queue.isEmpty()) {
            String nodeId = queue.removeFirst();
            GraphNode node = nodes.get(nodeId);
            if (node != null) {
                result.add(node);
            }
            List<String> neighbors = adjacency.getOrDefault(nodeId, new ArrayList<>());
            for (String neighbor : neighbors) {
                inDegree.merge(neighbor, -1, Integer::sum);
                if (inDegree.get(neighbor) == 0) {
                    queue.add(neighbor);
                }
            }
        }
        return result;
    }

    public void detectCycles() {
        List<GraphNode> sorted = topologicalSort();
        if (sorted.size() != nodes.size()) {
            throw new AlpError("Dependency cycle detected in ALP graph");
        }
    }

    public GraphNode getNode(String id) {
        return nodes.get(id);
    }
}

File: sdk/go/alp_graph.go
package alpgo

import (
	"fmt"
	"path/filepath"
	"sort"
)

type GraphNode struct {
	ID       string
	NodeType string
}

func NewGraphNode(id, nodeType string) *GraphNode {
	return &GraphNode{ID: id, NodeType: nodeType}
}

type AlpGraph struct {
	nodes     map[string]*GraphNode
	adjacency map[string][]string
}

func NewAlpGraph() *AlpGraph {
	return &AlpGraph{
		nodes:     make(map[string]*GraphNode),
		adjacency: make(map[string][]string),
	}
}

func (g *AlpGraph) BuildGraph(objects []*AlpObject) {
	g.nodes = make(map[string]*GraphNode)
	g.adjacency = make(map[string][]string)
	for _, obj := range objects {
		g.nodes[obj.ID] = NewGraphNode(obj.ID, obj.Type)
	}
	for _, obj := range objects {
		if depID, ok := obj.Properties["depends_on"].(string); ok {
			g.adjacency[depID] = append(g.adjacency[depID], obj.ID)
		}
	}
}

func (g *AlpGraph) TopologicalSort() []*GraphNode {
	inDegree := make(map[string]int)
	for nodeID := range g.nodes {
		inDegree[nodeID] = 0
	}
	for _, neighbors := range g.adjacency {
		for _, dep := range neighbors {
			inDegree[dep]++
		}
	}
	var queue []string
	for id, degree := range inDegree {
		if degree == 0 {
			queue = append(queue, id)
		}
	}
	sort.Strings(queue)
	var result []*GraphNode
	for len(queue) > 0 {
		nodeID := queue[0]
		queue = queue[1:]
		if node, ok := g.nodes[nodeID]; ok {
			result = append(result, node)
		}
		neighbors := g.adjacency[nodeID]
		var nextQueue []string
		for _, neighbor := range neighbors {
			inDegree[neighbor]--
			if inDegree[neighbor] == 0 {
				nextQueue = append(nextQueue, neighbor)
			}
		}
		sort.Strings(nextQueue)
		queue = append(queue, nextQueue...)
	}
	return result
}

func (g *AlpGraph) DetectCycles() {
	sorted := g.TopologicalSort()
	if len(sorted) != len(g.nodes) {
		panic("dependency cycle detected in ALP graph")
	}
}

func (g *AlpGraph) GetNode(id string) *GraphNode {
	return g.nodes[id]
}
